File: hex_terrain_editor/model/hex_cell.py
"""HexCell data model representing a single hex on the map."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from .terrain_types import Climate, Decorator, Elevation


@dataclass
class HexCell:
    """A single hex cell in the map grid."""

    q: int
    r: int
    elevation: Elevation = Elevation.PLAINS
    climate: Climate = Climate.Cf
    decorators: list[Decorator] = field(default_factory=list)
    label: str | None = None
    label_offset: tuple[int, int] = (0, 0)

    def copy(self) -> HexCell:
        """Return a deep copy of this cell."""
        return HexCell(
            q=self.q,
            r=self.r,
            elevation=self.elevation,
            climate=self.climate,
            decorators=list(self.decorators),
            label=self.label,
            label_offset=self.label_offset,
        )

    def to_dict(self) -> dict[str, Any]:
        """Serialize to a dictionary."""
        d: dict[str, Any] = {
            "q": self.q,
            "r": self.r,
            "elevation": self.elevation.name,
            "climate": self.climate.name,
        }
        if self.decorators:
            d["decorators"] = [dec.value for dec in self.decorators]
        if self.label:
            d["label"] = self.label
            d["label_offset"] = list(self.label_offset)
        return d

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> HexCell:
        """Deserialize from a dictionary."""
        decorators = [Decorator(v) for v in data.get("decorators", [])]
        label_offset = tuple(data.get("label_offset", [0, 0]))
        return cls(
            q=data["q"],
            r=data["r"],
            elevation=Elevation[data["elevation"]],
            climate=Climate[data["climate"]],
            decorators=decorators,
            label=data.get("label"),
            label_offset=(label_offset[0], label_offset[1]),
        )

    def add_decorator(self, decorator: Decorator) -> None:
        if decorator not in self.decorators:
            self.decorators.append(decorator)

    def remove_decorator(self, decorator: Decorator) -> None:
        if decorator in self.decorators:
            self.decorators.remove(decorator)

    @property
    def elevation_band(self) -> str:
        """Return the elevation band name for color lookup."""
        if self.elevation.is_water():
            return "water"
        elif self.elevation == Elevation.PLAINS:
            return "plains"
        elif self.elevation == Elevation.HILLS:
            return "hills"
        else:
            return "mountain"

File: src/model/hex_cell.rs
use serde::{Deserialize, Serialize};
use super::terrain_types::{Climate, Decorator, Elevation};

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct HexCell {
    pub q: i32,
    pub r: i32,
    pub elevation: Elevation,
    pub climate: Climate,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub decorators: Vec<Decorator>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub label: Option<String>,
    #[serde(default)]
    pub label_offset: (i32, i32),
}

impl HexCell {
    pub fn new(q: i32, r: i32) -> Self {
        Self {
            q,
            r,
            elevation: Elevation::Plains,
            climate: Climate::Cf,
            decorators: Vec::new(),
            label: None,
            label_offset: (0, 0),
        }
    }

    pub fn elevation_band(&self) -> &'static str {
        if self.elevation.is_water() {
            "water"
        } else {
            match self.elevation {
                Elevation::Hills => "hills",
                Elevation::Mountains => "mountain",
                _ => "plains",
            }
        }
    }

    pub fn add_decorator(&mut self, d: Decorator) {
        if !self.decorators.contains(&d) {
            self.decorators.push(d);
        }
    }

    pub fn remove_decorator(&mut self, d: Decorator) {
        self.decorators.retain(|existing| *existing != d);
    }

    pub fn texture_index(&self) -> u32 {
        self.elevation.texture_index(self.climate)
    }
}
